File: desafio-05/brunosana/javascript/desafio_05.js
const fs = require('fs');



//Questão 1
const global_max = function (array) {
    let array_max = [array[0]];
    array.forEach(f => {
        if (f.salario > array_max[0].salario){
            array_max = [];
            array_max.push(f);
        }else{
            if (f.salario == array_max[0].salario && f.id != array_max[0].id){
                array_max.push(f);
            }
        }
    });
    return array_max;
}
const global_min = function (array) {
    let array_min = [array[0]];
    array.forEach(f => {
        if (f.salario < array_min[0].salario){
            array_min = [];
            array_min.push(f);
        }else{
            if (f.salario == array_min[0].salario){
                array_min.push(f);
            }
        }
    });
    return array_min;
}
const global_avg = function(array){
    let avg = 0, n = 0;
    array.forEach(f => {
        avg += f.salario;
        n++;
    });
    return (avg/n).toFixed(2);
}

let funcionarios

fs.readFile(`./${process.argv[2]}`, (err, data) => {
    if (err) throw err;
    funcionarios = JSON.parse(data);
    //Questao 1_a
    global_min(funcionarios.funcionarios).forEach(f=> {
        console.log(`global_min | ${f.nome} ${f.sobrenome} | ${f.salario.toFixed(2)}`);
    })
    //Questao 1_b    
    global_max(funcionarios.funcionarios).forEach(f => {
        console.log(`global_max | ${f.nome} ${f.sobrenome} | ${f.salario.toFixed(2)}`);
    });
    
    //Questao 1 - c
    console.log(`global_avg | ${global_avg(funcionarios.funcionarios)}`);
    
    
    //Questão 2
    let areas_with_func_numbers = [];
    funcionarios.areas.forEach(e=>{
    let new_array = funcionarios.funcionarios.filter(f => f.area == e.codigo);
    //Questao 2_a
    if(new_array.length > 0){
        global_min(new_array).forEach(f=> {
            console.log(`area_min | ${e.nome} | ${f.nome} ${f.sobrenome} | ${f.salario.toFixed(2)}`);
        }) 
        //Questao 2_b
        global_max(new_array).forEach(f => {
            console.log(`area_max | ${e.nome} | ${f.nome} ${f.sobrenome} | ${f.salario.toFixed(2)}`);
        });
        
        console.log(`area_avg | ${e.nome} | ${global_avg(new_array)}`);
        e.numero_funcionarios = new_array.length;
        areas_with_func_numbers.push(e);
    }
    })
    
    areas_with_func_numbers.sort(function (ex, ey){
    if (ex.numero_funcionarios > ey.numero_funcionarios) return 1;
    })
    
    let initial_value = areas_with_func_numbers[0];
    areas_with_func_numbers.forEach( e => {
    initial_value.numero_funcionarios == e.numero_funcionarios ? console.log(`least_employeers | ${e.nome} | ${e.numero_funcionarios}`) : null;
    })
    initial_value = areas_with_func_numbers[areas_with_func_numbers.length-1];
    areas_with_func_numbers.forEach( e => {
    initial_value.numero_funcionarios == e.numero_funcionarios ? console.log(`most_employeers | ${e.nome} | ${e.numero_funcionarios}`) : null;
    })
    
    
    let sobrenomes = []
    funcionarios.funcionarios.forEach(e => {
    let have_sobrenome = false;
    sobrenomes.forEach(f => {
        if ((e.sobrenome == f) && !have_sobrenome){
            have_sobrenome = true;
        } 
    })
    have_sobrenome ? null : sobrenomes.push(e.sobrenome);
    })
    
    sobrenomes.forEach(e => {
    let vf = funcionarios.funcionarios.filter(avaliable = x => {
        return x.sobrenome == e;
    })
    if (vf.length > 1){
        global_max(vf).forEach(el => {
            console.log(`last_name_max | ${e} | ${el.nome} | ${el.salario}`);
        })
    }
    }) 
});

File: desafio-05/g4br13lr0ch4/python/desafio05.py
"""Solução do desafio 05 submetido por Gabriel Rocha - G4BR13LR0CH4"""

import json
import sys

FILE_NAME = sys.argv[1]

def question01(funcionarios):
    """
        Lista os funcionarios com os maiores e menores salarios,
        além de exibir o valor medio dos salarios
    """
    global_max = False
    global_min = False
    global_avg = False

    for funcionario in funcionarios:
        if global_min or global_max or global_avg:
            if funcionario['salario'] > global_max:
                global_max = funcionario['salario']

            if funcionario['salario'] < global_min:
                global_min = funcionario['salario']

            global_avg += funcionario['salario']

        else:
            global_min = global_avg = global_max = funcionario['salario']

    for funcionario in funcionarios:
        if funcionario['salario'] == global_max:
            print(f"global_max|{funcionario['nome']} {funcionario['sobrenome']}|" +
                  f"{funcionario['salario']:.2f}")

    for funcionario in funcionarios:
        if funcionario['salario'] == global_min:
            print(f"global_min|{funcionario['nome']} {funcionario['sobrenome']}|" +
                  f"{funcionario['salario']:.2f}")

    print(f"global_avg|{(global_avg/len(funcionarios)):.2f}")

def question02(file):
    """Separa quem mais recebe e quem menos recebe por area"""
    data = {}

    for funcionario in file['funcionarios']:
        if funcionario['area'] not in data:
            data[funcionario['area']] = [
                1,
                funcionario['salario'],
                funcionario['salario'],
                funcionario['salario']
            ]#quantidade de pessoas, salario maximo, salario minimo e medio.

        else:
            data[funcionario['area']][0] += 1 #quantidade de pessoas por area

            if data[funcionario['area']][1] < funcionario['salario']:
                data[funcionario['area']][1] = funcionario['salario'] #Maior salario

            if data[funcionario['area']][2] > funcionario['salario']:
                data[funcionario['area']][2] = funcionario['salario'] #Menor salario

            data[funcionario['area']][3] += funcionario['salario'] #Salario medio

    for area in file['areas']:
        if area['codigo'] not in data:
            data[area['codigo']] = [0, 0, 0, 0]

    for area in file['areas']:
        for funcionario in file['funcionarios']:
            if (
                    (funcionario['area'] == area['codigo']) and
                    (funcionario['salario'] == data[area['codigo']][1])
            ):
                print(f"area_max|{area['nome']}|{funcionario['nome']} {funcionario['sobrenome']}|" +
                      f"{funcionario['salario']:.2f}")

        for funcionario in file['funcionarios']:
            if (
                    (funcionario['area'] == area['codigo']) and
                    (funcionario['salario'] == data[area['codigo']][2])
            ):
                print(f"area_min|{area['nome']}|{funcionario['nome']} " +
                      f"{funcionario['sobrenome']}|{funcionario['salario']:.2f}")

        if data[area['codigo']][0] > 0:
            print(f"area_avg|{area['nome']}|" +
                  f"{((data[area['codigo']][3])/data[area['codigo']][0]):.2f}")

    return data

def question03(areas, data):
    """Informa a area com mais pessoas funcionarios e a com menor"""
    most_employees = 0
    least_employees = 0

    for area in areas:
        if most_employees == 0 and least_employees == 0:
            least_employees = most_employees = data[area['codigo']][0]
            most = {
                "area": area['nome'],
                "quantidade": data[area['codigo']][0]
            }
            least = {
                "area": area['nome'],
                "quantidade": data[area['codigo']][0]
            }

        if most_employees < data[area['codigo']][0]:
            most_employees = data[area['codigo']][0]
            most = {"area": area['nome'], "quantidade": data[area['codigo']][0]}

        if least_employees > data[area['codigo']][0] and data[area['codigo']][0] != 0:
            least_employees = data[area['codigo']][0]
            least = {"area": area['nome'], "quantidade": data[area['codigo']][0]}

    print(f"most_employees|{most['area']}|{most['quantidade']}")
    print(f"least_employees|{least['area']}|{least['quantidade']}")

def question04(funcionarios):
    """Separa quem ganha mais comparando entre os sobrenomes"""
    data = {}

    for funcionario in funcionarios:
        if funcionario['sobrenome'] not in data:
            data[funcionario['sobrenome']] = [1, funcionario['salario']]
        else:
            data[funcionario['sobrenome']][0] += 1
            if data[funcionario['sobrenome']][1] < funcionario['salario']:
                data[funcionario['sobrenome']][1] = funcionario['salario']

    for funcionario in funcionarios:
        if (
                (data[funcionario['sobrenome']][0] > 1) and
                (funcionario['salario'] == data[funcionario['sobrenome']][1])
        ):
            print(f"last_name_max|{funcionario['sobrenome']}|" +
                  f"{funcionario['nome']} {funcionario['sobrenome']}|{funcionario['salario']:.2f}")

def load():
    """Carrega o arquivo e decodifica"""
    with open(FILE_NAME, 'r') as json_file:
        file = json.load(json_file)
        question01(file['funcionarios'])
        question03(file['areas'], question02(file))
        question04(file['funcionarios'])

if __name__ == "__main__":
    load()
